src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,644 @@
+    }
+}
+
+pub struct BTreeEngineIterator {
+    tree: Arc<RwLock<BTreeMap<Key, Value>>>,
+    cur_key: Option<Key>,
+    cur_value: Option<Value>,
+    valid: bool,
+    lower_bound: Bound<Key>,
+    upper_bound: Bound<Key>,
+}
+
+impl BTreeEngineIterator {
+    pub fn new(
+        tree: Arc<RwLock<BTreeMap<Key, Value>>>,
+        iter_opt: IterOption,
+    ) -> BTreeEngineIterator {
+        let lower_bound = match iter_opt.lower_bound() {
+            None => Unbounded,
+            Some(key) => Included(Key::from_raw(key)),
+        };
+
+        let upper_bound = match iter_opt.upper_bound() {
+            None => Unbounded,
+            Some(key) => Excluded(Key::from_raw(key)),
+        };
+        assert!(
+            check_bounds(&lower_bound, &upper_bound),
+            "lower bound must <= upper bound,lower_bound: {:?}, upper_bound:{:?}",
+            lower_bound,
+            upper_bound
+        );
+
+        Self {
+            tree: tree.clone(),
+            cur_key: None,
+            cur_value: None,
+            valid: false,
+            lower_bound,
+            upper_bound,
+        }
+    }
+}
+
+impl Iterator for BTreeEngineIterator {
+    fn next(&mut self) -> bool {
+        let tree = self.tree.read().unwrap();
+        let mut range = tree.range((
+            Excluded(self.cur_key.clone().unwrap()),
+            self.upper_bound.clone(),
+        ));
+        match range.next() {
+            None => {
+                self.valid = false;
+            }
+            Some((k, v)) => {
+                self.cur_key = Some(k.clone());
+                self.cur_value = Some(v.clone());
+                self.valid = true;
+            }
+        }
+        self.valid()
+    }
+
+    fn prev(&mut self) -> bool {
+        let tree = self.tree.read().unwrap();
+        let mut range = tree.range((
+            self.lower_bound.clone(),
+            Excluded(self.cur_key.clone().unwrap()),
+        ));
+        match range.next_back() {
+            None => {
+                self.valid = false;
+            }
+            Some((k, v)) => {
+                self.cur_key = Some(k.clone());
+                self.cur_value = Some(v.clone());
+                self.valid = true;
+            }
+        }
+        self.valid()
+    }
+
+    fn seek(&mut self, key: &Key) -> EngineResult<bool> {
+        let key_bound = Included(key.clone());
+        let upper_bound = self.upper_bound.clone();
+        assert!(
+            check_key_in_range(&self.lower_bound, &key_bound, &upper_bound),
+            "Err(KeyNotInRegion),start_key: {:?}, end_key:{:?}",
+            key_bound,
+            upper_bound
+        );
+
+        debug!("Iterator.seek({:?})", unsafe {
+            String::from_utf8_unchecked(key.to_raw().unwrap())
+        });
+        match self
+            .tree
+            .read()
+            .unwrap()
+            .range((key_bound, upper_bound))
+            .next()
+        {
+            None => {
```

Comment:
I think there's duplicate logic (see line 169 to 177), can we extract a function?

Reply:
Good idea, I am working on it.